DCOS-48316: Jobs form

natmegs opened the pull request:
Implements a basic new Jobs form. The modal and form are based on the Services implementations (`CreateServiceModal`, `CreateServiceModalForm`). 

Included in these commits:
- Basic tabbed job form (that can toggle to JSON editor) and modal
- General form section component
- Modifications to `MetronomeClient` to upgrade to metronome API v1 spec for job create and update
- Job editing

Not included in this PR:
- The form doesn't open to it's own dedicated route (do we want this? @AmrAbdelrazik)

This PR overlaps a couple different JIRA tickets. These can be separated into separate PRs if necessary but it's difficult to test the changes end to end without all 3:

### DCOS-48316: Jobs form modal
- Adds a modal and form component
- Create/edit buttons open a full page modal
- Form contains one tab (General)
- Basic validation and error visibility is set up
- Modal can toggle a JSON editor

### DCOS-48445: Datalayer changes
- Updates the create and update job mutations to be compliant with the metronome v1 API spec (previously was v0)
- See https://github.com/dcos/dcos-ui/pull/3599/commits/6ac68557ae19eb7b71c9aedbf4c485ec22d75ddb, https://github.com/dcos/dcos-ui/pull/3599/commits/30a00897fb573ae6479e1f5292f2f16116627751, https://github.com/dcos/dcos-ui/pull/3599/commits/5a05f2c8c63dc7e7caf751c28270207a1b1dae46

### DCOS-48352: General form section
- Adds the general form section to enter the bare minimum for a job (ID, description, cpus, mem, disk, gpus, command, container Image)
- Should be able to edit from both form inputs and JSON
- Validation should prevent submission of an invalid form and show errors

## Testing

Different scenarios to try:
- Create a job (it should work)
- When 'Command Only' is chosen, only Command input should be shown below (required)
- When 'Container Image' is chosen, both Container Image (required) and Command inputs should be shown below (also, a `ucr` property should be added to the job JSON - visible in JSON editor)
- Try putting the validation to the test (try submitting an invalid form)
- Try editing using the JSON editor
- Try creating a job and adding properties to the JSON that aren't yet available in the form UI. These should persist when the job is added. For example, try adding a schedule like:
```
{
"job": {...},
"schedule": {
    "id": "my-schedule-data",
    "cron": "0 4 * * *"
  }
}
```


## Trade-offs

This was put together using existing patterns from the Services form. The tradeoff here is using what is tried & tested & available (fast but comes with its own challenges and potential tech debt) vs coming up with novel form patterns.

## Dependencies



## Screenshots

<img width="1212" alt="screen shot 2019-02-14 at 5 16 51 pm" src="https://user-images.githubusercontent.com/19582796/52828129-70ea6c00-307c-11e9-89fa-85a2d876b317.png">
<img width="1210" alt="screen shot 2019-02-14 at 5 17 02 pm" src="https://user-images.githubusercontent.com/19582796/52828139-78117a00-307c-11e9-847c-4d597cde6628.png">
<img width="1210" alt="screen shot 2019-02-14 at 5 17 18 pm" src="https://user-images.githubusercontent.com/19582796/52828153-7fd11e80-307c-11e9-96ae-f8903ef27783.png">

AmrAbdelrazik commented:
not sure I understand what you mean by: "The form doesn't open to it's own dedicated route (do we want this?)"

natmegs commented:
The service form opens in its own route (`/services/overview/create`) whereas the jobs form just opens as an overlay on the `/jobs/overview` page

AmrAbdelrazik commented:
I don't have any strong opinions here, I am ok with the same behavior like Marathon Form

Poltergeist commented on plugins/jobs/src/js/components/form/GeneralFormSection.tsx:
```diff
@@ -0,0 +1,365 @@
+            min="32"
+            name="job.run.mem"
+            type="number"
+            value={job.run.mem}
```
Something we should do for all nested fields is we should use [findNestedPropertyInObject](https://github.com/dcos/dcos-ui/blob/master/src/js/utils/Util.js#L79-L91) We can't be sure that the fields are present end if `run` is not present the whole form will break.

GeorgiSTodorov commented:
Looks good. Some observations:
- When I open the new form, the Disk input is immediately glowing red as if I did something wrong.
![screenshot from 2019-02-20 10-58-35](https://user-images.githubusercontent.com/40791275/53079250-b1c00600-34fe-11e9-883d-6e6ef7694a44.png)
- I think that "Container Image" and "Command Only" radio buttons would look better if they were aligned with the other inputs
- I can't write in the "Description" input (is this known?)
![screenshot from 2019-02-20 11-15-15](https://user-images.githubusercontent.com/40791275/53080208-d0270100-3500-11e9-9407-01a0a775dee4.png)
- I can't submit the form, because it requires docker image even though I selected "Command only"
![screenshot from 2019-02-20 11-04-25](https://user-images.githubusercontent.com/40791275/53079548-5a6e6580-34ff-11e9-91ba-11452283fe8c.png)
- I get "Object is not valid" for Disk and GPU even though I entered integer values (Disk needs to be above 32, but this isn't obvious)
![screenshot from 2019-02-20 11-05-52](https://user-images.githubusercontent.com/40791275/53079657-90abe500-34ff-11e9-88a4-8ff1954532ef.png)
- All tabs expect "General" don't work (this might be known)
- I think we don't need a tooltip explanation
![screenshot from 2019-02-20 11-09-37](https://user-images.githubusercontent.com/40791275/53079913-25164780-3500-11e9-830a-226349a2eaf3.png)
- No message catalog changes (this is probably known)
- I don't get an error message when I enter an uppercase job name
- I am not sure if we need both a description and a tooltip for command

natmegs commented:
@GeorgiSTodorov Thanks for the detailed feedback! Looks like the last commit regressed a lot of the validation behaviour - I'll check it out.

natmegs commented:
@GeorgiSTodorov I pushed the following changes to address your feedback:
- fixed validation issues
- removed unused tabs (only general left)
- added message catalog updates
- added validation for job id
- re: tooltips, I removed the tooltips on Mem and Disk but to stay true to the design I've kept the tooltips on ID, description, CPUs, Command and Container Image (as well as added one to GPUs with a message clarifying it is only available with UCR) (cc @izharbuendia)
Let me know if any of the issues persist for you

AmrAbdelrazik commented:
I actually liked the tooltips, but I will leave it to @izharbuendia

On Wed, Feb 20, 2019 at 5:57 PM Natalie Smith <notifications@github.com>
wrote:

> @GeorgiSTodorov <https://github.com/GeorgiSTodorov> I pushed the
> following changes to address your feedback:
>
>    - fixed validation issues
>    - removed unused tabs (only general left)
>    - added message catalog updates
>    - added validation for job id
>    - re: tooltips, I removed the tooltips on Mem and Disk but to stay
>    true to the design I've kept the tooltips on ID, description, CPUs, Command
>    and Container Image (as well as added one to GPUs with a message clarifying
>    it is only available with UCR) (cc @izharbuendia
>    <https://github.com/izharbuendia>)
>    Let me know if any of the issues persist for you
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dcos/dcos-ui/pull/3599#issuecomment-465830344>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AOMJeQz2rCBBKfPyRxDo5-oKYHuCFHDQks5vPf0GgaJpZM4a8wLK>
> .
>
-- 
Amr Abdelrazik
Director of Product Marketing
amr@mesosphere.com

Poltergeist commented:
Some issues I was running into while testing: 
![2019-02-21 16 49 21](https://user-images.githubusercontent.com/156010/53182214-5292ec00-35f9-11e9-999c-26b27c3f8ccf.gif)

The JSON editor produces a non-estimated output when changing a key field as seen in the JSON.

The other issue is that the JSON editor is reseting every couple of seconds, if I haven't reached a valid new state.

mperrotti commented on plugins/jobs/src/js/components/form/GeneralFormSection.tsx:
```diff
@@ -0,0 +1,351 @@
+              </FormGroupHeadingContent>
+            </FormGroupHeading>
+          </FieldLabel>
+          <FieldInput min="0.01" name="job.run.cpus" value={formData.cpus} />
```
Should this be `type="number"`?

mperrotti commented:
@izharbuendia - we should come up with some guidelines on when to use a tooltip for form hint text versus adding the info as hint text below the input.
For example: the info in the "Description" tooltip is very short and simple, but it's still hidden in a tooltip instead of just being hint text below the input

natmegs commented:
Updates from today:
- parsing number values for cpus, mem, disk, gpus
- allow multiple consecutive edits of a job
- amend job id regex check to allow single character id
- make invalid json errors from editor block form submission
- json editor should no longer "reset" when in an invalid state
- allow deletion of schedule from a job in json editor
- remove min from disk (it was wrong anyway and our validation checks for it separately)
- add type="number" to cpus
- fix update logic in MetronomeClient so that if a job is being updated and a user adds a schedule where the original job previously didn't have one, it will POST the new schedule instead of trying to PUT it and erroring
- change what happens when a job is being created and the JOB creation is successful but the SCHEDULE creation fails: show error in modal, switch to edit state, allow user to modify and resubmit as an update request (cc @AmrAbdelrazik )

@GeorgiSTodorov @mperrotti I was unable to reproduce the 500 error, let me know if it persists. Also, we are not implementing the review screen in this RI so that's why the button says "Submit". Let me know if I've missed anything!

GeorgiSTodorov commented:
@natmegs , thank you. Most of my concerns are addressed. Some minor things:
- CPU and GPU inputs glow red as soon as the form is opened
- CPU is red after entering invalid value and changing it to valid
- GPU is red when it is below 32 and cannot be changed below 32 using the arrows (is this intended?).
- When I am editing the form and I enter a new value for cpu, disk, mem and I click submit, the old values are visible for a split second. This might be misleading, because the user might see this and think that his new values weren't saved.
- I am able to enter "abc" value for GPU and the submit succeeds and "abc" is simply not saved (GPU is saved as 0).
- When CPU is for example 10, clicking the up arrow changes it to 10.01 and clicking the down arrow changes it to 9.01. This is valid according to the tooltip, but still seems weird to me.

Poltergeist commented on plugins/jobs/src/js/components/JobsForm.tsx:
```diff
@@ -0,0 +1,214 @@
+    );
+  }
+
+  onInputChange(event: any) {
```
Let's use the proper type for the event here https://github.com/sw-yx/react-typescript-cheatsheet#forms-and-events.

Poltergeist commented on plugins/jobs/src/js/components/JobsForm.tsx:
```diff
@@ -0,0 +1,214 @@
+    const { onChange } = this.props;
+    const newValue = event.target.value;
+    const inputName = event.target.name;
+    const inputType = event.target.type;
```
How do we feel about chaning the top three lines  to 
```js
const {value, name, type} = event.target;

const action = {
  value,
  path: name,
  type: type === "number"
    ? JobFormActionType.SetNum
     : JobFormActionType.Set
};
```

natmegs commented:
@GeorgiSTodorov I've addressed your comments with a couple limitations:
- GPUs: in Chrome, I cannot enter a string into the number input, and if I enter `gpus: "abc"` in the json editor it throws an error on submit (I think this is how it should work). In Firefox, I can enter a string in the number input (yikes, what's the point of a number input in that case) but it won't add it to the JSON (I think this is also behaving as it should given the browser limitation).
- Firefox does some really weird stuff with html form validation so I put `novalidate` on the form and removed the min restraints on the inputs (we check for these in our own validation). CPUs will now increment by 1 using arrows (and I changed default to 1 as indicated in designs) but user can still enter 1.01 (in firefox this sometimes causes the input to highlight red for unknown reasons but it still is added to the json). Now when I open the form, nothing glows red in either Chrome or Firefox, let me know if this is different for you.

GeorgiSTodorov commented:
@natmegs , looks good.
- I think that the id field should be disabled when editing the job
- When I delete a value in the JSON (for example CPU) and click Submit, the old value is back, but the error won't go away and I can't submit the form even when changing the value
![peek 2019-02-25 11-35](https://user-images.githubusercontent.com/40791275/53328045-c54de100-38f1-11e9-9b03-b089bca71122.gif)
- If I open JSON editor, save the job and then edit it, JSON editor isn't open (it is in the old form and I think it makes sense to be)
- `npm run commitlint` is failing
- If I enter a negative value for GPUs, it says "Object is not valid"

natmegs commented:
@GeorgiSTodorov Some good points here but not all are things we can address in this PR as they are working as designed or are separate features.
- Disable ID on edit: I agree with this one! But that's not how our jobs or services forms work currently so we can extract that into a separate ticket if it's a product priority (cc @AmrAbdelrazik )
- Invalid JSON issue: Good catch, I've pushed a fix so that the JSON editor won't get overwritten (https://cl.ly/cee893e9a70c)
- Open JSON editor on edit if it was last used when creating the job: if I'm understanding correctly, this is a user preference that we have no way of storing for jobs
- The 'object is not valid' message for negative gpus is from the server (since it's not valid 🙂). This is expected and if you change it to a valid value and submit it will work. I've changed the error message extraction to retrieve the detailed server error message if it exists rather than the generic one (here's an example of a server error being surfaced https://cl.ly/bd2f1e18d9ba). As additional security I also added a validation check for gpus > 0 (https://cl.ly/09701ab28e85).

@mperrotti @juliangieseke let me know if your concerns have been addressed :) Mike, if it's ok I'd like to upgrade to the ui-kit modal in a separate ticket since it will likely break integration/system tests and I don't have a reliable way to run system tests locally right now.

GeorgiSTodorov commented:
> Disable ID on edit: I agree with this one! But that's not how our jobs or services forms work currently so we can extract that into a separate ticket if it's a product priority (cc @AmrAbdelrazik )
We disable it for services.

![screenshot from 2019-02-26 10-26-44](https://user-images.githubusercontent.com/40791275/53398079-6e0f4580-39b1-11e9-9569-3bc1d9ed4d7d.png)
And there is a bug about it not being disabled for jobs - https://jira.mesosphere.com/browse/DCOS-39292 .

> Open JSON editor on edit if it was last used when creating the job: if I'm understanding correctly, this is a user preference that we have no way of storing for jobs

It works in the old form:
![peek 2019-02-26 10-31](https://user-images.githubusercontent.com/40791275/53398208-b9295880-39b1-11e9-899f-c3de83662cde.gif)

GeorgiSTodorov commented:
I pushed a commit that should fix the system tests.

natmegs commented:
@GeorgiSTodorov thanks so much for fixing those system tests! Very appreciated. Those are all good points but have been determined to be non-blockers for this particular PR:
- Disabled ID: since it's not disabled in the current jobs form, and we are trying to reach feature parity with the existing jobs form, this can be fixed separately (thanks for linking the ticket!)
- JSON editor state: that's actually a bug in the current jobs form where the state isn't being reset correctly 🙂 
- 3 errors: that is actually by design. The user has 3 options for how to remedy that error
- Browser width: that is how the services form looks at that browser width, and mobile browser layouts are not a priority for this particular PR
Hope this addresses your concerns! Happy to talk about them further if there are questions.

Poltergeist commented:
@GeorgiSTodorov 

- The Id field disable we might not be doing in this round since we are aiming for feature parity first.
- That the JSON editor is active seems to me actually more of a bug than an actual feature. It seems we forgot to reset the JSON editor flag.  
> It's not a Bug it's a Feature. – Someone from the internetz 
- About the 3 errors for one field 
- The responsive design is okay, it is the same problem for the service form.

We will fix those issues in a future PR and they should not be blocking this PR from moving forward.

Poltergeist commented on Jenkinsfile:
```diff
@@ -55,7 +55,7 @@ pipeline {
 
       steps {
-        sh 'npm run commitlint -- --from "${CHANGE_TARGET}"'
+        sh 'npm run commitlint -- --from origin/"${CHANGE_TARGET}"'
```
This we need for the feature branch workflow to work.

GeorgiSTodorov commented:
On first glance, it looks good, I will do an extensive review tomorrow.

GeorgiSTodorov commented:
These are probably not blocking changes, but I still think they would be nice to haves:
- As @mperrotti said, the "Submit" button is supposed to be "Review"
Design document: https://mesosphere.invisionapp.com/share/Y6Q8EFKBA3G#/screens/341644097
- Also in the design doc, GPUs is not disabled if we have selected "Command only"
- If I try to submit a "command only" job without a command, I get an error. When I switch to "container image", the command is no longer required, but still shows an error. The error goes away if I enter a container image
- I still think we need the little tooltips that say "Description info", "CPUs info"
- If I enter a super high value in CPUs, it says `Minimum value is 0.1.` and the JSON for cpu is an empty string (same for other inputs)
- Console warnings when I open the modal
![screenshot from 2019-02-28 10-07-13](https://user-images.githubusercontent.com/40791275/53551020-dd676f80-3b40-11e9-9c27-dc0b8abc6b17.png)

GeorgiSTodorov commented on plugins/jobs/src/js/components/form/helpers/MetronomeJobValidators.ts:
```diff
@@ -0,0 +1,258 @@
+   * Returns whether a field is shown based on current state of form data.
+   */
+  fieldIsShown(fieldName: string, formData: FormOutput): boolean {
+    switch (fieldName) {
```
It's the same in both cases. Is this expected?

GeorgiSTodorov commented:
@mperrotti

pierrebeitz commented:
i removed the newly introduced console-warnings and all type errors that would still be there after https://github.com/dcos/dcos-ui/pull/3623 is merged.

GeorgiSTodorov commented:
How do we push commits without removing the approval? Is it because it's not pointed to master?

nLight commented:
@GeorgiSTodorov right, target branch is not protected

GeorgiSTodorov commented:
@nLight , is this intentional?

nLight commented:
@GeorgiSTodorov we protect only master and release branches. We can install a protection for that branch if needed. Everyone should have admin access to do it

mperrotti commented on plugins/jobs/src/js/components/JobsFormModal.tsx:
```diff
@@ -0,0 +1,355 @@
+import { graphqlObservable } from "@dcos/data-service";
+import { take } from "rxjs/operators";
+
+import FullScreenModal from "#SRC/js/components/modals/FullScreenModal";
```
@natmegs - I guess this doesn't have to be a blocker, but I'd prefer we use ui-kit components when possible so we can deprecate `reactjs-components` and `CNVS` sooner than later

natmegs replied:
@mperrotti I'll replace it as part of my spike today once this merges into the feature branch!

juliangieseke commented:
…and CI is red :(
